src/FSharp.Control.AsyncSeq/AsyncSeq.fs

```diff
@@ -986,6 +986,44 @@ module AsyncSeq =
       if (buffer.Count > 0) then 
           yield buffer.ToArray() }
 
+  let bufferByCountAndTime (bufferSize:int) (timeoutMs:int) (source:AsyncSeq<'T>) : AsyncSeq<'T[]> = 
+    if (bufferSize < 1) then invalidArg "bufferSize" "must be positive"
+    if (timeoutMs < 1) then invalidArg "timeoutMs" "must be positive"
+    asyncSeq {
+      let buffer = new ResizeArray<_>()
+      use ie = source.GetEnumerator()
+      let rec loop rem rt = asyncSeq {
```

Comment:
Most of the rest of the code uses "while" loops, for better or worse.  These use a Landin's-knot like trick to amortize the allocation that controls the loop. You may want to compare perf between these approaches at some point.